src/stacked_borrows/diagnostics.rs

```diff
@@ -104,77 +346,141 @@ impl AllocHistory {
     }
 
     /// Report a descriptive error when `new` could not be granted from `derived_from`.
-    pub fn grant_error<'tcx>(
-        &self,
-        derived_from: ProvenanceExtra,
-        new: Item,
-        alloc_id: AllocId,
-        alloc_range: AllocRange,
-        error_offset: Size,
-        stack: &Stack,
-    ) -> InterpError<'tcx> {
+    pub fn grant_error(&self, perm: Permission, stack: &Stack) -> InterpError<'tcx> {
+        let Operation::Retag(op) = &self.operation else {
+            unreachable!("grant_error should only be called during a retag")
+        };
         let action = format!(
-            "trying to reborrow from {derived_from:?} for {new_perm:?} permission at {alloc_id:?}[{offset:#x}]",
-            new_perm = new.perm(),
-            offset = error_offset.bytes(),
+            "trying to retag from {:?} for {:?} permission at {:?}[{:#x}]",
+            op.orig_tag,
+            perm,
+            self.history.id,
+            self.offset.bytes(),
         );
         err_sb_ub(
-            format!("{}{}", action, error_cause(stack, derived_from)),
-            Some(operation_summary("a reborrow", alloc_id, alloc_range)),
-            derived_from.and_then(|derived_from| self.get_logs_relevant_to(derived_from, None)),
+            format!("{}{}", action, error_cause(stack, op.orig_tag)),
+            Some(operation_summary(&op.cause.summary(), self.history.id, op.range)),
+            op.orig_tag.and_then(|orig_tag| self.get_logs_relevant_to(orig_tag, None)),
         )
     }
 
     /// Report a descriptive error when `access` is not permitted based on `tag`.
-    pub fn access_error<'tcx>(
-        &self,
-        access: AccessKind,
-        tag: ProvenanceExtra,
-        alloc_id: AllocId,
-        alloc_range: AllocRange,
-        error_offset: Size,
-        stack: &Stack,
-    ) -> InterpError<'tcx> {
+    pub fn access_error(&self, stack: &Stack) -> InterpError<'tcx> {
+        let Operation::Access(op) = &self.operation  else {
+            unreachable!("access_error should only be called during an access")
+        };
         let action = format!(
             "attempting a {access} using {tag:?} at {alloc_id:?}[{offset:#x}]",
-            offset = error_offset.bytes(),
+            access = op.kind,
+            tag = op.tag,
+            alloc_id = self.history.id,
+            offset = self.offset.bytes(),
         );
         err_sb_ub(
-            format!("{}{}", action, error_cause(stack, tag)),
-            Some(operation_summary("an access", alloc_id, alloc_range)),
-            tag.and_then(|tag| self.get_logs_relevant_to(tag, None)),
+            format!("{}{}", action, error_cause(stack, op.tag)),
+            Some(operation_summary("an access", self.history.id, op.range)),
+            op.tag.and_then(|tag| self.get_logs_relevant_to(tag, None)),
         )
     }
+
+    pub fn protector_error(&self, item: &Item) -> InterpError<'tcx> {
```

Comment:
This lost its `inline(never)`.

Should some other functions in this file have that, too?